### DISCUSSION TITLE
Should performance optimizations be paywalled?

### DISCUSSION THREAD

**Discussion by AoElite** [redisson/redisson#5724] [Polls]
I can get the different cache implementations & other features, but why are actual performance optimizations like the "Ultra-fast Redis client engine" as advertised on the website being paywalled? It makes it seem like the open-source version is purposely being slowed or less optimized.

**Comment by mrniko** [comment]
> It makes it seem like the open-source version is purposely being slowed or less optimized.

That's an interesting statement. Could you please show me that code to prove that? If so Jedis and Lettuce did the same, since they have almost the same speed as Redisson open-source version.

**Reply by AoElite** [reply]
Redisson Pro is marketed as being up to four times faster than its open-source counterpart. This claim suggests that the performance of the open-source version may be deliberately limited. While this might not be seen as malicious or intentional sabotage, it does raise questions about the extent of optimization withheld from the free version.

**Reply by mrniko** [reply]
Again, according to your deductions Jedis and Lettuce are also deliberately limited, since they have almost the same speed as Redisson open-source version.

PRO is faster than Jedis also in almost 4 times https://dzone.com/articles/redisson-pro-vs-jedis-which-is-faster 

PRO version has a completely different engine. It's not just a couple optimizations. As the author, and I reserve the right to decide which version it should be used in. 

I'd be interested in discussing all these questions with someone who has an experience of development of their own open-source project with commercial version. Working full-time without attracting investment or working anywhere else, at the same time making a commercial model that would ensure its sustainable development. Otherwise, discussing what can be paid and what can't be paid makes no sense. Sorry.

**Reply by AoElite** [reply]
Jedis & Lettuce don't have pro versions, so they obviously aren't purposely limiting their performance, unlike Redisson.

**Reply by mrniko** [reply]
Like I said before, it's not purposely limited it's just has an engine with the same performance capabilities as Jedis & Lettuce.

**Reply by AoElite** [reply]
Redisson isn't Jedis or Lettuce. You can compare their performance all you want, but that doesn't change the fact that the paid version of Redisson is 4x faster than the open-source version.

**Reply by rodrigoGA** [reply]
I understand the need for commercial viability, nobody works for free, and I’m grateful that an open-source edition of Redisson exists. That said, as an engineer choosing a client today, the performance split between the OSS and Pro versions makes me hesitant to adopt Redisson. It creates a feeling of low confidence around the OSS version. I totally get charging for advanced features, but paywalling core performance nudges me toward Jedis instead.

For reference, I found another comparison that reports similar behavior with Redisson when using local caching (a capability Jedis already supports): https://www.instaclustr.com/blog/redis-java-clients-and-client-side-caching/  Unfortunately, I haven’t come across any up to date head tohead benchmarks. If there are recent, transparent OSS benchmarks (and maybe a roadmap for OSS performance), that would really help build confidence.

**Reply by AoElite** [reply]
Yea I agree, totally get advanced features being paid but not performance optimizations.

**Reply by mrniko** [reply]
@rodrigoGA 

> It creates a feeling of low confidence around the OSS version. I totally get charging for advanced features, but paywalling core performance nudges me toward Jedis instead.

Jedis and Redisson OS has almost similar performance. As for PRO version results are here https://dzone.com/articles/redisson-pro-vs-jedis-which-is-faster

**Reply by AoElite** [reply]
Well that is an extremely outdated benchmark so I wouldn't use that as confirmation as to which is faster in it's current state.

**Reply by adrian-skybaker** [reply]
>That said, as an engineer choosing a client today, the performance split between the OSS and Pro versions makes me hesitant to adopt Redisson. It creates a feeling of low confidence around the OSS version

Having used the OSS version only for 5+ years in some fairly highly tuned deployments, in practice I've not seen this as an issue. 

Some specific observations:
- The presence of the commercial version has given us more confidence in the OSS version, as it has meant more support, activity and investment in the project that flows through to the OSS version.
- The OSS version is simply not slow. We've never seen the engine itself surface as a measureable overhead compared to network latency and serialization/deserialization overhead.
- The aspects of the PRO version that at times have appealed to us have not been general "optimisations", but additional features for specific scenarios (for example, some of the local cache variants and reliable messaging implementation)
- We've not tested it, but if you look at the whitepaper for the "ultra fast" mode, my impression is that you would likely only benefit from this in some scenarios, and be needing to be upwards of 30-40k operations per sec.

**Reply by rodrigoGA** [reply]
I built a small open benchmark to compare Jedis and Redisson (OSS), with and without client-side caching. (Honesty note: most of the harness was AI-generated; the configs are simple and reproducible.)

**What I saw:**
- Without local cache, **Jedis** generally delivered better or comparable throughput and **consistent** reads across workloads.
- With local cache enabled, **Redisson** showed **higher throughput** on some workloads, **but** also **frequent read inconsistencies (stale reads)** in read-mostly tests—enough that I wouldn’t choose it for scenarios that need predictable read-your-writes.
- **Jedis with local cache** did **not** show those inconsistencies in my runs.

If useful to others, the repo (scripts + results + CSV summaries) is here: https://github.com/rodrigoGA/Redis-Client-Benchmarks-Jedis-vs-Redisson

**Reply by mrniko** [reply]
> Without local cache, Jedis generally delivered better or comparable throughput and consistent reads across workloads.

Can you share the exact numbers?

**Reply by rodrigoGA** [reply]
here https://github.com/rodrigoGA/Redis-Client-Benchmarks-Jedis-vs-Redisson/blob/master/benchmark-results/multi-run-summary.csv
